src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -78,80 +81,120 @@ public void onRefreshSettings(Settings settings) {
+                        foundTranslogs.add(new FsChannelImmutableReader(id, raf, raf.channel().size(), FsChannelReader.UNKNOWN));
+                        logger.debug("found local translog with id [{}]", id);
+                    }
+                }
+            }
+            CollectionUtil.timSort(foundTranslogs, new Comparator<FsChannelReader>() {
+                @Override
+                public int compare(FsChannelReader o1, FsChannelReader o2) {
+                    return Long.compare(o1.translogId(), o2.translogId());
+                }
+            });
+            uncomittedTranslogs.addAll(foundTranslogs);
+        }
+    }
 
-        this.type = FsTranslogFile.Type.fromString(indexSettings.get("index.translog.fs.type", FsTranslogFile.Type.BUFFERED.name()));
-        this.bufferSize = (int) indexSettings.getAsBytesSize("index.translog.fs.buffer_size", ByteSizeValue.parseBytesSizeValue("64k")).bytes();
+    protected long parseIdFromFileName(Path translogFile) {
+        final String fileName = translogFile.getFileName().toString();
+        final Matcher matcher = PARSE_ID_PATTERN.matcher(fileName);
+        if (matcher.matches()) {
+            try {
+                return Long.parseLong(matcher.group(1));
+            } catch (NumberFormatException e) {
+                // will be logged bellow
+            }
+        }
+        logger.trace("failed to extract translog id from [{}]", fileName);
+        return -1;
     }
 
     @Override
     public void updateBuffer(ByteSizeValue bufferSize) {
         this.bufferSize = bufferSize.bytesAsInt();
-        rwl.writeLock().lock();
-        try {
-            FsTranslogFile current1 = this.current;
-            if (current1 != null) {
-                current1.updateBufferSize(this.bufferSize);
-            }
-            current1 = this.trans;
-            if (current1 != null) {
-                current1.updateBufferSize(this.bufferSize);
-            }
-        } finally {
-            rwl.writeLock().unlock();
+        try (ReleasableLock _ = writeLock.acquire()) {
+            current.updateBufferSize(this.bufferSize);
         }
     }
 
+    public boolean isOpen() {
+        return closed.get() == false;
+    }
+
     @Override
     public void close() throws IOException {
-        if (indexSettingsService != null) {
-            indexSettingsService.removeListener(applySettings);
-        }
-        rwl.writeLock().lock();
-        try {
-            IOUtils.close(this.trans, this.current);
-        } finally {
-            rwl.writeLock().unlock();
+        if (closed.compareAndSet(false, true)) {
+            if (indexSettingsService != null) {
+                indexSettingsService.removeListener(applySettings);
+            }
+            try (ReleasableLock _ = writeLock.acquire()) {
+                IOUtils.close(this.current);
+                IOUtils.close(uncomittedTranslogs);
```

Comment:
You should be able to collapse this to `IOUtils.close(this.current, uncomittedTranslogs)`